### PR TITLE
feat: adding support for using multiple keychain for sending sbom results to a different repository

### LINKED ISSUE #819: let putting SBOMs in a different repository
We can enable putting SBOMs in a different repository by creating another environment variable or a new configuration option.

```sh
# via environment variable
SBOM_REPOSITORY=<repo> ko build ..

# via the configuration file
.ko.yaml:
sbomRepository: <repo> 
```

cc: @dentrax

> 🗣 Based on the talk: https://kubernetes.slack.com/archives/C032MM2CH7X/p1662936574448189
> 🧵 Cross-ref: https://github.com/kyverno/kyverno/issues/4592

### PR DESCRIPTION
**Pull request by developer-guy**
Signed-off-by: Batuhan Apaydın <batuhan.apaydin@trendyol.com>

Fixes #819

### CODE REVIEW

File: pkg/publish/options.go [hunk 1]
```diff
@@ -42,38 +38,11 @@ func WithUserAgent(ua string) Option {
...
-	return func(i *defaultOpener) error {
-		i.auth = auth
-		return nil
-	}
-}
-
-// WithAuthFromKeychain is a functional option for overriding the default
+// WithKeyChain is a functional option for overriding the default
 // authenticator on a default publisher using an authn.Keychain
-func WithAuthFromKeychain(keys authn.Keychain) Option {
```
Comment (imjasonh) [inline pkg/publish/options.go:56]: We shouldn't rename this, since other users who use ko as a library might use it. Same with `WithAuth` above.

File: pkg/publish/default.go [hunk 2]
```diff
@@ -129,7 +133,7 @@ func pushResult(ctx context.Context, tag name.Tag, br build.Result, opt []remote
 		ociOpts := []ociremote.Option{ociremote.WithRemoteOptions(opt...)}
 
 		// Respect COSIGN_REPOSITORY
-		targetRepoOverride, err := ociremote.GetEnvTargetRepository()
```
Comment (imjasonh) [inline pkg/publish/default.go:132]: Does this mean we _already_ support `COSIGN_REPOSITORY`? 🤯 

If so, maybe we should just use this for now and document a plan to support non-`COSIGN`-named env vars in the future.

Comment (developer-guy) [reply pkg/publish/default.go:132]: >Does this mean we already support COSIGN_REPOSITORY? 🤯

Haha, yes 🫣

>If so, maybe we should just use this for now and document a plan to support non-COSIGN-named env vars in the future.

As you wish, but this keychain problem remains the same, at least we should fix this.

Comment (imjasonh) [reply pkg/publish/default.go:132]: Ah I get what you're saying now. Yeah, instead of taking an auth we should take a keychain, since we might be pushing to two repos/registries due to COSIGN_REPOSITORY.

Thanks for trying this out and for figuring out the fix. 👍

Comment (developer-guy) [reply pkg/publish/default.go]: >instead of taking an auth we should take a keychain

Exactly!

File: pkg/publish/default.go [hunk 3]
```diff
@@ -67,8 +69,9 @@ type Namer func(string, string) string
 
 // identity is the default namer, so import paths are affixed as-is under the repository
 // name for maximum clarity, e.g.
-//   gcr.io/foo/github.com/bar/baz/cmd/blah
-//   ^--base--^ ^-------import path-------^
+
+//	gcr.io/foo/github.com/bar/baz/cmd/blah
+//	^--base--^ ^-------import path-------^
```
Comment (imjasonh) [inline pkg/publish/default.go:74]: undo this change.

File: pkg/commands/resolver.go [hunk 4]
```diff
@@ -209,7 +209,7 @@ func makePublisher(po *options.PublishOptions) (publish.Interface, error) {
 		if po.Push {
 			dp, err := publish.NewDefault(repoName,
 				publish.WithUserAgent(userAgent),
-				publish.WithAuthFromKeychain(keychain),
+				publish.WithKeyChain(keychain),
```
Comment (imjasonh) [inline pkg/commands/resolver.go:212]: This has to keep the name `publish.WithAuthFromKeychain`.

File: pkg/publish/options.go [hunk 5]
```diff
@@ -78,6 +78,15 @@ func WithAuthFromKeychain(keys authn.Keychain) Option {
 	}
 }
 
+// WithKeyChain is a functional option for overriding the default
+// authenticator on a default publisher using an authn.Keychain
+func WithKeyChain(keys authn.Keychain) Option {
```
Comment (imjasonh) [inline pkg/publish/options.go:83]: We don't need to add a new method, we can just keep using `WithAuthFromKeychain`.

**Review by imjasonh — COMMENTED** [review]
Lookin' good, just a few small things.

File: pkg/publish/default.go [hunk 6]
```diff
@@ -203,7 +207,7 @@ func (d *defalt) Publish(ctx context.Context, br build.Result, s string) (name.R
 	// https://github.com/google/go-containerregistry/issues/212
 	s = strings.ToLower(s)
 
-	ro := []remote.Option{remote.WithAuth(d.auth), remote.WithTransport(d.t), remote.WithContext(ctx), remote.WithUserAgent(d.userAgent)}
+	ro := []remote.Option{remote.WithAuthFromKeychain(d.keychain), remote.WithTransport(d.t), remote.WithContext(ctx), remote.WithUserAgent(d.userAgent)}
```
Comment (imjasonh) [inline pkg/publish/default.go:210]: Hmm... So now we have the case where a client may have used `publish.WithAuth`, and with this change we may be dropping it on the floor.

I think what we should do is check whether `d.auth == authn.Anonymous`, and only if it is, add `remote.WithAuthFromKeychain(d.keychain)` -- otherwise, keep `remote.WithAuth(d.auth)`. We might want to error if both are passed too, to avoid confusion.

This should only affect clients like Skaffold that use this code as a library though.

cc @halvards

Comment (developer-guy) [reply pkg/publish/default.go]: I fixed it, it should be okay right now.

Comment (halvards) [reply pkg/publish/default.go]: Thanks for keeping ko-as-library users in mind :slightly_smiling_face:

Comment (developer-guy) [reply pkg/publish/default.go]: kindly ping @imjasonh ☝️

File: pkg/publish/default.go [hunk 7]
```diff
@@ -203,7 +207,13 @@ func (d *defalt) Publish(ctx context.Context, br build.Result, s string) (name.R
 	// https://github.com/google/go-containerregistry/issues/212
 	s = strings.ToLower(s)
 
-	ro := []remote.Option{remote.WithAuth(d.auth), remote.WithTransport(d.t), remote.WithContext(ctx), remote.WithUserAgent(d.userAgent)}
+	ro := []remote.Option{remote.WithTransport(d.t), remote.WithContext(ctx), remote.WithUserAgent(d.userAgent)}
+	if d.auth == authn.Anonymous {
```
Comment (imjasonh) [inline pkg/publish/default.go:211]:
```suggestion
	if d.auth != authn.Anonymous {
```

Right?

Maybe we should only store a `keychain`, and if someone calls `WithAuth` we should set `d.keychain` to a keychain that only returns that auth. Then we don't have to handle a possible situation where we store both an auth and a keychain, and have to decide which one to use.

Comment (developer-guy) [reply pkg/publish/default.go:211]: >Maybe we should only store a keychain
That's how I did it first. Do you want me to revert to the first version of the development?

^Based on that, if I understand correctly, we should remove the if logic and use the keychain only.

Right?

Comment (imjasonh) [reply pkg/publish/default.go:211]: Yeah, this struct would only hold a `keychain`, either the exactly the one passed by `WithAuthFromKeychain`, or a new keychain impl that always returns the auth passed by `WithAuth`.

Comment (developer-guy) [reply pkg/publish/default.go:211]: >this struct would only hold a keychain, either the exactly the one passed by WithAuthFromKeychain

I would like to go with this, wdyt?

**Comment by realshuting** [comment]
Thanks for the work @developer-guy and the review @imjasonh ! Would like to get the support in!

**Comment by developer-guy** [comment]
I think we are in good shape about this PR

File: pkg/publish/default.go [hunk 8]
```diff
@@ -40,8 +40,9 @@ type defalt struct {
 	base      string
 	t         http.RoundTripper
 	userAgent string
-	auth      authn.Authenticator
 	namer     Namer
+	auth      authn.Authenticator
```
Comment (imjasonh) [inline pkg/publish/default.go:44]: This shouldn't have an `auth` field anymore, only `keychain`. `publish.WithAuth` should set `keychain = staticKeychain{auth}`.

File: pkg/publish/default.go [hunk 9]
```diff
@@ -55,6 +56,7 @@ type defaultOpener struct {
 	t         http.RoundTripper
 	userAgent string
 	auth      authn.Authenticator
```
Comment (imjasonh) [inline pkg/publish/default.go:58]: Same here.

File: pkg/publish/default.go [hunk 10]
```diff
@@ -203,7 +208,13 @@ func (d *defalt) Publish(ctx context.Context, br build.Result, s string) (name.R
 	// https://github.com/google/go-containerregistry/issues/212
 	s = strings.ToLower(s)
 
-	ro := []remote.Option{remote.WithAuth(d.auth), remote.WithTransport(d.t), remote.WithContext(ctx), remote.WithUserAgent(d.userAgent)}
+	ro := []remote.Option{remote.WithTransport(d.t), remote.WithContext(ctx), remote.WithUserAgent(d.userAgent)}
+	if d.auth != authn.Anonymous {
```
Comment (imjasonh) [inline pkg/publish/default.go:212]: This should just always pass `remote.WithAuthFromKeychain(d.keychain)`, which is `staticKeychain{authn.Anonymous}` by default.

File: pkg/publish/options.go [hunk 11]
```diff
@@ -16,14 +16,19 @@ package publish
...
 
 	"github.com/google/go-containerregistry/pkg/authn"
-	"github.com/google/go-containerregistry/pkg/name"
 )
 
+type staticKeychain struct {
+	auth authn.Authenticator
+}
+
+func (s *staticKeychain) Resolve(resource authn.Resource) (authn.Authenticator, error) {
```
Comment (imjasonh) [inline pkg/publish/options.go:28]: nit: just `staticKeychain` (no `*`), since there's no situation where a call should update the value.

File: pkg/publish/options.go [hunk 12]
```diff
@@ -55,25 +60,7 @@ func WithAuth(auth authn.Authenticator) Option {
...
-			return err
-		}
-		auth, err := keys.Resolve(repo.Registry)
-		if err != nil {
-			return err
-		}
-		if auth == authn.Anonymous {
-			log.Println("No matching credentials were found, falling back on anonymous")
-		}
-		i.auth = auth
```
[the remark this answers is not in the document]
Comment (developer-guy) [reply pkg/publish/options.go:76]: >will they still see this line?

I think they won't see this line anymore since we are not resolving any keychain to auth. Also, shouldn't it be an expected thing thus, we are changing the auth strategy by looking at the keychain itself with this change. What is your suggestion?

**Review by imjasonh — APPROVED** [review]

**Comment by imjasonh** [comment]
I'd love to see a test for this new behavior, but I've already taken long enough reviewing this so I won't block on it.

Thanks for adding this! 👍
